core/src/vaults.ts

```diff
@@ -0,0 +1,446 @@
+    return {
+        ...vaultBase,
+        ...vaultCollateralParameters,
+        ...vaultAmount,
+        ...globalLiquidationLimits,
+        currentCollateralDebtDai,
+        maximumCollateralDebtDai,
+        liquidationPenaltyRatio,
+        minimalAuctionedDai,
+    };
+};
+export const fetchVault = memoizee(_fetchVault, {
+    maxAge: CACHE_EXPIRY_MS,
+    promise: true,
+    length: 2,
+});
+
+const getNextOraclePrice = async (
+    oracle: CollateralPriceSourceConfig,
+    provider: ethers.providers.JsonRpcProvider,
+    oracleAddress: string
+): Promise<BigNumber> => {
+    if (!('nextPriceSlotAddress' in oracle)) {
+        return new BigNumber(NaN);
+    }
+    const slot = oracle.nextPriceSlotAddress;
+    const nextPriceFeed = await provider.getStorageAt(oracleAddress, slot);
+    const valueSplitPosition = oracle.slotPriceValueBeginsAtPosition;
+    const isPriceValid = nextPriceFeed.substring(0, valueSplitPosition);
+    if (parseInt(isPriceValid, 16) === 1) {
+        return new BigNumber(`0x${nextPriceFeed.substring(valueSplitPosition)}`);
+    }
+    return new BigNumber(NaN);
+};
+
+const getCurrentOraclePrice = async (
+    oracle: CollateralPriceSourceConfig,
+    provider: ethers.providers.JsonRpcProvider,
+    oracleAddress: string
+) => {
+    const currentPriceFeed = await provider.getStorageAt(oracleAddress, oracle.currentPriceSlotAddress);
+    const valueSplitPosition = oracle.slotPriceValueBeginsAtPosition;
+    let isPriceValid;
+    if ('currentPriceValiditySlotAndOffset' in oracle) {
+        isPriceValid = await provider.getStorageAt(oracleAddress, oracle.currentPriceValiditySlotAndOffset.slot);
+        isPriceValid = isPriceValid[oracle.currentPriceValiditySlotAndOffset.offset];
+    } else {
+        isPriceValid = currentPriceFeed.substring(0, valueSplitPosition);
+    }
+    return parseInt(isPriceValid, 16) === 1
+        ? new BigNumber(`0x${currentPriceFeed.substring(valueSplitPosition)}`)
+        : new BigNumber(NaN);
+};
+
+const getNextOraclePriceChange = async (
+    collateralConfig: CollateralConfig,
+    provider: ethers.providers.JsonRpcProvider,
+    oracleAddress: string
+) => {
+    let nextPriceChange: Date = new Date(NaN);
+    if (collateralConfig.oracle.hasDelay) {
+        const osmContractInterface = await getContractInterfaceByName('OSM');
+        const osmContract = new ethers.Contract(oracleAddress, osmContractInterface, provider);
+        const lastPriceUpdateAsHex = (await osmContract.zzz())._hex;
+        const priceUpdateFrequencyInSeconds = await osmContract.hop();
+        const lastPriceUpdateTimestampInSeconds = new BigNumber(lastPriceUpdateAsHex).toNumber();
+        nextPriceChange = new Date((lastPriceUpdateTimestampInSeconds + priceUpdateFrequencyInSeconds) * 1000);
+    }
+    return nextPriceChange;
+};
+
+const _getOsmPrices = async (
+    network: string,
+    oracleAddress: string,
+    collateralType: CollateralType
+): Promise<OraclePrices> => {
```

Comment:
Let's remove oracle logic into a separate file, it's too long and complicated and surely deserves a better structure